File: boxnote2md/runner.py
"""ファイル/ディレクトリ単位の変換オーケストレータ。"""

from __future__ import annotations

import sys
from dataclasses import dataclass, field
from pathlib import Path

from boxnote2md.reader import BoxNoteParseError, read_boxnote
from boxnote2md.render import RenderContext, render_document
from boxnote2md.writer import derive_output_path, resolve_collision, write_markdown


@dataclass
class RunOptions:
    input_path: Path
    out_dir: Path
    recursive: bool = True
    flat: bool = False
    overwrite: bool = False
    image_mode: str = "download"
    image_dir: Path | None = None
    keep_styles: bool = False
    dry_run: bool = False
    verbose: bool = False


@dataclass
class RunSummary:
    success: int = 0
    skipped: int = 0
    failed: int = 0
    failures: list[tuple[Path, str]] = field(default_factory=list)


def _log(msg: str, *, verbose: bool = False, ctx_verbose: bool = False) -> None:
    if verbose and not ctx_verbose:
        return
    print(msg, file=sys.stderr)


def _collect_inputs(opts: RunOptions) -> tuple[list[Path], Path]:
    """変換対象ファイルのリストと、出力パス計算用のルートを返す。"""
    inp = opts.input_path
    if inp.is_file():
        return [inp], inp
    if inp.is_dir():
        files = sorted(inp.rglob("*.boxnote") if opts.recursive else inp.glob("*.boxnote"))
        return files, inp
    raise FileNotFoundError(f"input not found or unsupported: {inp}")


def _convert_one(src: Path, dest: Path, opts: RunOptions) -> None:
    doc = read_boxnote(src)
    ctx = RenderContext(
        keep_styles=opts.keep_styles,
        image_mode=opts.image_mode,
        image_dir=opts.image_dir,
        md_path=dest,
    )
    md = render_document(doc, ctx)
    if opts.verbose:
        for w in ctx.warnings:
            _log(f"  warn: {w}", verbose=True, ctx_verbose=True)
    status = write_markdown(dest, md, overwrite=opts.overwrite, dry_run=opts.dry_run)
    if opts.verbose or opts.dry_run:
        _log(f"  -> {dest} [{status}]", verbose=True, ctx_verbose=True)


def run(opts: RunOptions) -> RunSummary:
    summary = RunSummary()

    try:
        files, input_root = _collect_inputs(opts)
    except FileNotFoundError as e:
        _log(f"error: {e}", verbose=True, ctx_verbose=True)
        summary.failed = 1
        summary.failures.append((opts.input_path, str(e)))
        return summary

    if not files:
        _log(
            f"no .boxnote files found under {opts.input_path}",
            verbose=True,
            ctx_verbose=True,
        )
        return summary

    if opts.verbose:
        _log(f"found {len(files)} file(s)", verbose=True, ctx_verbose=True)

    taken: set[Path] = set()
    for src in files:
        dest = derive_output_path(
            src,
            input_root=input_root,
            out_dir=opts.out_dir,
            flat=opts.flat,
        )
        dest = resolve_collision(dest, taken)
        taken.add(dest)

        if opts.verbose:
            _log(f"converting {src}", verbose=True, ctx_verbose=True)

        try:
            existed_before = dest.exists()
            _convert_one(src, dest, opts)
        except BoxNoteParseError as e:
            summary.failed += 1
            summary.failures.append((src, str(e)))
            _log(f"FAILED: {src}: {e}", verbose=True, ctx_verbose=True)
            continue
        except Exception as e:  # noqa: BLE001
            summary.failed += 1
            summary.failures.append((src, repr(e)))
            _log(f"FAILED: {src}: {e}", verbose=True, ctx_verbose=True)
            continue

        # skipped 判定: 既存があり overwrite=False
        if existed_before and not opts.overwrite:
            summary.skipped += 1
        else:
            summary.success += 1

    print(
        f"summary: success={summary.success} skipped={summary.skipped} failed={summary.failed}",
        file=sys.stderr,
    )
    if summary.failures:
        print("failures:", file=sys.stderr)
        for path, msg in summary.failures:
            print(f"  - {path}: {msg}", file=sys.stderr)

    return summary

File: internal/runner/runner.go
// Package runner は CLI のファイル/ディレクトリ走査と変換実行をまとめる。
package runner

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/nanakanok/boxnote2md-cli/internal/boxnote"
	"github.com/nanakanok/boxnote2md-cli/internal/mdparse"
	"github.com/nanakanok/boxnote2md-cli/internal/render"
)

// BoxnoteToMdOptions は boxnote2md コマンドのオプション。
type BoxnoteToMdOptions struct {
	Input      string
	OutDir     string
	Recursive  bool
	Flat       bool
	Overwrite  bool
	ImageMode  string
	ImageDir   string
	KeepStyles bool
	DryRun     bool
	Verbose    bool
}

// MdToBoxnoteOptions は md2boxnote コマンドのオプション。
type MdToBoxnoteOptions struct {
	Input     string
	OutDir    string
	Recursive bool
	Flat      bool
	Overwrite bool
	DryRun    bool
	Verbose   bool
}

// Summary は変換結果のサマリ。
type Summary struct {
	Success  int
	Skipped  int
	Failed   int
	Failures []FailureRecord
}

// FailureRecord は失敗 1 件分の記録。
type FailureRecord struct {
	Path string
	Err  error
}

// RunBoxnoteToMd は .boxnote → .md 変換を実行する。
func RunBoxnoteToMd(opts BoxnoteToMdOptions) (Summary, error) {
	imageDir := opts.ImageDir
	if imageDir == "" {
		imageDir = filepath.Join(opts.OutDir, "images")
	}

	files, root, err := collectInputs(opts.Input, ".boxnote", opts.Recursive)
	if err != nil {
		return Summary{Failed: 1, Failures: []FailureRecord{{Path: opts.Input, Err: err}}}, nil
	}
	if opts.Verbose {
		logf("found %d file(s)", len(files))
	}

	var summary Summary
	taken := map[string]struct{}{}

	for _, src := range files {
		dest := deriveOutPath(src, root, opts.OutDir, ".md", opts.Flat)
		dest = resolveCollision(dest, taken)
		taken[dest] = struct{}{}

		if opts.Verbose {
			logf("converting %s", src)
		}

		existed := fileExists(dest)
		err := convertBoxnoteToMd(src, dest, imageDir, opts)
		if err != nil {
			summary.Failed++
			summary.Failures = append(summary.Failures, FailureRecord{Path: src, Err: err})
			logf("FAILED: %s: %v", src, err)
			continue
		}
		if existed && !opts.Overwrite {
			summary.Skipped++
		} else {
			summary.Success++
		}
	}

	printSummary(summary)
	return summary, nil
}

func convertBoxnoteToMd(src, dest, imageDir string, opts BoxnoteToMdOptions) error {
	if fileExists(dest) && !opts.Overwrite {
		if opts.Verbose || opts.DryRun {
			logf("  -> %s [skipped]", dest)
		}
		return nil
	}
	doc, err := boxnote.Read(src)
	if err != nil {
		return err
	}
	ctx := &render.Context{
		KeepStyles: opts.KeepStyles,
		ImageMode:  opts.ImageMode,
		ImageDir:   imageDir,
		MdPath:     dest,
	}
	md, err := render.Document(doc, ctx)
	if err != nil {
		return err
	}
	if opts.Verbose {
		for _, w := range ctx.Warnings {
			logf("  warn: %s", w)
		}
	}
	if opts.DryRun {
		logf("  -> %s [would-write]", dest)
		return nil
	}
	if err := os.MkdirAll(filepath.Dir(dest), 0o755); err != nil {
		return err
	}
	if err := os.WriteFile(dest, []byte(md), 0o644); err != nil {
		return err
	}
	if opts.Verbose {
		logf("  -> %s [written]", dest)
	}
	return nil
}

// RunMdToBoxnote は .md → .boxnote 変換を実行する。
func RunMdToBoxnote(opts MdToBoxnoteOptions) (Summary, error) {
	files, root, err := collectInputs(opts.Input, ".md", opts.Recursive)
	if err != nil {
		return Summary{Failed: 1, Failures: []FailureRecord{{Path: opts.Input, Err: err}}}, nil
	}
	if opts.Verbose {
		logf("found %d file(s)", len(files))
	}

	var summary Summary
	taken := map[string]struct{}{}
	for _, src := range files {
		dest := deriveOutPath(src, root, opts.OutDir, ".boxnote", opts.Flat)
		dest = resolveCollision(dest, taken)
		taken[dest] = struct{}{}

		if opts.Verbose {
			logf("converting %s", src)
		}

		existed := fileExists(dest)
		err := convertMdToBoxnote(src, dest, opts)
		if err != nil {
			summary.Failed++
			summary.Failures = append(summary.Failures, FailureRecord{Path: src, Err: err})
			logf("FAILED: %s: %v", src, err)
			continue
		}
		if existed && !opts.Overwrite {
			summary.Skipped++
		} else {
			summary.Success++
		}
	}

	printSummary(summary)
	return summary, nil
}

func convertMdToBoxnote(src, dest string, opts MdToBoxnoteOptions) error {
	if fileExists(dest) && !opts.Overwrite {
		if opts.Verbose || opts.DryRun {
			logf("  -> %s [skipped]", dest)
		}
		return nil
	}
	raw, err := os.ReadFile(src)
	if err != nil {
		return err
	}
	doc := mdparse.Parse(string(raw))
	env := mdparse.WrapEnvelope(doc)

	if opts.DryRun {
		logf("  -> %s [would-write]", dest)
		return nil
	}
	if err := os.MkdirAll(filepath.Dir(dest), 0o755); err != nil {
		return err
	}
	out, err := json.Marshal(env)
	if err != nil {
		return err
	}
	if err := os.WriteFile(dest, out, 0o644); err != nil {
		return err
	}
	if opts.Verbose {
		logf("  -> %s [written]", dest)
	}
	return nil
}

// ============================================================
// 共通ユーティリティ
// ============================================================

func collectInputs(input, ext string, recursive bool) ([]string, string, error) {
	info, err := os.Stat(input)
	if err != nil {
		return nil, "", fmt.Errorf("input not found or unsupported: %s: %w", input, err)
	}
	if !info.IsDir() {
		return []string{input}, input, nil
	}
	var files []string
	if recursive {
		err = filepath.WalkDir(input, func(path string, d fs.DirEntry, err error) error {
			if err != nil {
				return err
			}
			if !d.IsDir() && strings.EqualFold(filepath.Ext(path), ext) {
				files = append(files, path)
			}
			return nil
		})
		if err != nil {
			return nil, "", err
		}
	} else {
		entries, err := os.ReadDir(input)
		if err != nil {
			return nil, "", err
		}
		for _, e := range entries {
			if !e.IsDir() && strings.EqualFold(filepath.Ext(e.Name()), ext) {
				files = append(files, filepath.Join(input, e.Name()))
			}
		}
	}
	sort.Strings(files)
	return files, input, nil
}

func deriveOutPath(src, root, outDir, newExt string, flat bool) string {
	base := strings.TrimSuffix(filepath.Base(src), filepath.Ext(src)) + newExt
	if flat || src == root {
		return filepath.Join(outDir, base)
	}
	rel, err := filepath.Rel(root, src)
	if err != nil {
		return filepath.Join(outDir, base)
	}
	rel = strings.TrimSuffix(rel, filepath.Ext(rel)) + newExt
	return filepath.Join(outDir, rel)
}

func resolveCollision(path string, taken map[string]struct{}) string {
	if _, exists := taken[path]; !exists {
		return path
	}
	dir := filepath.Dir(path)
	ext := filepath.Ext(path)
	stem := strings.TrimSuffix(filepath.Base(path), ext)
	for i := 1; ; i++ {
		cand := filepath.Join(dir, fmt.Sprintf("%s-%d%s", stem, i, ext))
		if _, exists := taken[cand]; !exists {
			return cand
		}
	}
}

func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

func logf(format string, args ...any) {
	fmt.Fprintf(os.Stderr, format+"\n", args...)
}

func printSummary(s Summary) {
	logf("summary: success=%d skipped=%d failed=%d", s.Success, s.Skipped, s.Failed)
	if len(s.Failures) > 0 {
		logf("failures:")
		for _, f := range s.Failures {
			logf("  - %s: %v", f.Path, f.Err)
		}
	}
}

// ensure errors.New is used so the package compiles even when error path changes.
var _ = errors.New
